code/renderers/vue3/src/render.ts

```diff
@@ -22,65 +23,81 @@ export const setup = (fn: (app: any) => void) => {
 
 const map = new Map<
   VueRenderer['canvasElement'],
-  { vueApp: ReturnType<typeof createApp>; reactiveArgs: any }
+  {
+    vueApp: ReturnType<typeof createApp>;
+    reactiveArgs: Args;
+  }
 >();
-
-const elementMap = new Map<VueRenderer['canvasElement'], StoryFnVueReturnType>();
-
+let reactiveState: {
+  globals: Globals;
+};
 export function renderToCanvas(
-  { storyFn, forceRemount, showMain, showException, storyContext }: RenderContext<VueRenderer>,
+  { storyFn, forceRemount, showMain, showException, storyContext, id }: RenderContext<VueRenderer>,
   canvasElement: VueRenderer['canvasElement']
 ) {
-  // fetch the story with the updated context (with reactive args)
-  storyContext.args = reactive(storyContext.args);
-  const element: StoryFnVueReturnType = storyFn();
-  elementMap.set(canvasElement, element);
-
-  const props = (element as any).render?.().props;
-  const reactiveArgs = props ? reactive(props) : storyContext.args;
-
   const existingApp = map.get(canvasElement);
+
+  // if the story is already rendered and we are not forcing a remount, we just update the reactive args
   if (existingApp && !forceRemount) {
-    updateArgs(existingApp.reactiveArgs, reactiveArgs);
+    const element = storyFn();
+
+    updateArgs(existingApp.reactiveArgs, element.props ?? storyContext.args);
     return () => {
       teardown(existingApp.vueApp, canvasElement);
     };
   }
-
   if (existingApp && forceRemount) teardown(existingApp.vueApp, canvasElement);
 
-  const storybookApp = createApp({
-    render() {
-      const renderedElement: any = elementMap.get(canvasElement);
-      const current = renderedElement && renderedElement.template ? renderedElement : element;
-      map.set(canvasElement, { vueApp: storybookApp, reactiveArgs });
-      return h(current, reactiveArgs);
+  // create vue app for the story
+  const vueApp = createApp({
+    setup() {
+      reactiveState = reactive({ globals: storyContext.globals });
+      storyContext.args = reactive(storyContext.args);
+      let rootElement = storyFn();
+      const appState = {
+        vueApp,
+        reactiveArgs: reactive(rootElement.props ?? storyContext.args),
+      };
+      map.set(canvasElement, appState);
+
+      watch(
+        () => reactiveState.globals,
+        () => {
+          storyContext.globals = reactiveState.globals;
+          rootElement = storyFn();
```

Comment:
Ideally, we should try a solution where is not needed right?

Reply:
Yes But for now Tom is saying we stick with react until we found a solution.